Ready according to previous review - extracted all the needed functions from base-private

Tcharl opened the pull request:
Fix remarks from https://github.com/jhipster/generator-jhipster/pull/20151

Contributes to https://github.com/jhipster/generator-jhipster/issues/20295
---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

<!--
Please also reference the issue number in a commit message to [automatically close the related GitHub issue](https://help.github.com/articles/closing-issues-via-commit-messages/)

Note: It is also possible to add `[skip ci]` or `[ci skip]` to your commit message to skip continuous integration tests
-->

mshima commented on generators/base/generator-base-private.mjs:
```diff
@@ -166,11 +162,8 @@ export default class PrivateBase extends Generator {
-      shelljs.rm('-rf', folder);
-    }
+    // TODO Should not update variable
+    folder = deleteFolder(this, folder);
```
I don’t think extracting a simple implementation like this into a non reusable lib is better.
IMO makes much more complicated.

Tcharl replied:
At least it highlights all the bad practices that were used ^^ (note that assignment is mandatory to get it working here :-S).

The goal behind this is to extract  the logic outside the generator to be able to rewrite the API in a much simpler way: there is a lot of duplicated and reusable code hidden within these 'simple implementations'

mshima replied:
I agree to extract logic outside the generator, but in this case removeFolder IMO is a generator basic operation.
It's basically the same that yeoman-generator `removeDestination` provides with the difference that this one removes from the physical disk, while the yeoman implementation removes from mem-fs and will be committed latter.

Tcharl replied:
This serie of PR will never end and the result won't be that much readable if the guideline is not a binary one (logic == `logic` folder, generator file == public API for blueprint).

mshima commented on generators/base/generator-base-private.mjs:
```diff
@@ -183,13 +175,10 @@ export default class PrivateBase extends Generator {
    * @returns {boolean} true if success; false otherwise
    */
   gitMove(source, dest) {
```
We probably should remove gitMove.
It was used to rename i18n files that change only the casing, breaking case insensitive filesystem.
Add an anonymous function at https://github.com/jhipster/generator-jhipster/blob/c4223686a9457e99c50a41584b950f3cafa8e23c/generators/cleanup.mjs#L362.
So we can decide if we can drop this upgrade logic later.

Tcharl replied:
I guess that there are a lot of file move (for example during jhipster upgrade).
We'll choose the best one once every filesystem move will be centralized.

mshima replied:
We don't do git operations at main generators.
We could do at upgrade generator, move there then.

Tcharl replied:
Sure, fully agree! Just want to be sure that the gitMove is only used this way.
 
Would you agree with this procedure:
1. extract the logic in the relevant's blueprint 'logic' folder in a serie of PR
2. check the usage of this business logic and move the generator method in the respective generator in another serie of PR
3. check if there is value exposing the generator method and remove it from the API if there isn't.

This approach looks straightforward, dumb enough to be reviewed quickly (and it should be because there will be a LOT of PRs to achieve before the 8th release) and less error prone than mixing purposes

Would it be ok to proceed this way?
I'll create the 3 different GH issue and reference the according PRs

Tcharl replied:
Done sir

Tcharl replied:
rollbacked for further investigation later on

mshima commented on generators/base/generator-base.mjs:
```diff
@@ -255,11 +253,7 @@ export default class JHipsterBaseGenerator extends PrivateBase {
    * @return {boolean} true if the User is built-in.
    */
   isUsingBuiltInUser() {
```
It's used only at https://github.com/jhipster/generator-jhipster/blob/c4223686a9457e99c50a41584b950f3cafa8e23c/generators/server/generator.mjs#L302.
Remove this api and move to server generator.

Tcharl replied:
Not sure that it's server only: there may be some impact on the client too (for example generating the registration form)

mshima replied:
Maybe you are confusing the derived property `application.builtInUser` with the application implementation.
Look close at the logic:
```
 return (
     !config ||
     (!config.skipUserManagement && config.databaseType !== NO_DATABASE) ||
     (config.authenticationType === OAUTH2 && config.databaseType !== NO_DATABASE)
   );
```

The server defines the `application.builtInUser` value and client generator uses it.
So in this case the logic is server only, and the client uses it.
Our client definition just mirrors what the server implements.

Tcharl replied:
No confusing with derived props, just challenging why this derived property is related to `server`. To me it should be an `entities` global logic and derived property assignment should be there.

Tcharl replied:
We can discuss it in front of a virtual beer ^^. Microservice+Jwt and Gw+Oauth is not a `server` concern, but an `application` one.
I may be confused, but to me `server` is related to one service's backend, for example the backend of one microservice or the backend of one monolith, which is why I didn't want to put that method here.
Am I wrong?

As an illustration: I've no use case in mind where the `isBuiltInUser` method will return `true`at the `gateway`level while it will return `false`at the `microservice` level within the same `application`.

mshima commented on generators/base/generator-base.mjs:
```diff
@@ -269,7 +263,7 @@ export default class JHipsterBaseGenerator extends PrivateBase {
    * @return {boolean} true if the entity is User.
    */
   isUserEntity(entityName) {
```
Looks like this can be dropped, no uses.

Tcharl replied:
I guess it will be in some templates, kind of derived entity property

mshima replied:
I can be wrong, but I think I've already replaced every use of this api.

Tcharl replied:
Used indirectly in client/.../entity file update

Tcharl replied:
Done sir

Tcharl replied:
dropped :-)

mshima commented on generators/base/generator-base.mjs:
```diff
@@ -288,11 +282,7 @@ export default class JHipsterBaseGenerator extends PrivateBase {
    * @return {boolean} true if the Authority is built-in.
    */
   isUsingBuiltInAuthority() {
```
Only used https://github.com/jhipster/generator-jhipster/blob/c4223686a9457e99c50a41584b950f3cafa8e23c/generators/server/generator.mjs#L303.
Same as the isUsingBuiltInUser

Tcharl replied:
Which is important ^^

Tcharl replied:
Not as easy at it seems:
Ideally it should be moved to `base-app` generator right? BUT it's used by the `isBuiltInAuthority(entityName)` of the `entity` generator (and in some templates), which doesn't inherit from this one (it inherits form base).
I agree that it should be done, but generator inheritance should be thought and reworked as a prerequisite (And I definitely don't feel confident to achieve it by myself)

Tcharl replied:
After a lot of try, I won't try to move the methods from generator to others before moving the code: there are too much interleaving and fully externalizing first will definitely help

mshima replied:
Let me do it. I’ve already started.
We will need to split the user/userManagement logic.

mshima commented on generators/base/generator-base.mjs:
```diff
@@ -302,7 +292,7 @@ export default class JHipsterBaseGenerator extends PrivateBase {
    * @return {boolean} true if the entity is Authority.
    */
   isAuthorityEntity(entityName) {
```
can be dropped.

Tcharl replied:
done

mshima commented on generators/base/generator-base.mjs:
```diff
@@ -346,25 +322,21 @@ export default class JHipsterBaseGenerator extends PrivateBase {
    * @param {string} packageJsonSourceFile - Package json filepath with actual versions.
    */
   replacePackageJsonVersions(keyToReplace, packageJsonSourceFile) {
```
can be dropped

mshima replied:
Another problem of all those methods is easy to just forget about them.
The utilities created when split this method should be removed too.

Tcharl replied:
removed sir

mshima commented on generators/base/generator-base-private.mjs:
```diff
@@ -153,11 +151,8 @@ export default class PrivateBase extends Generator {
    * @param file
    */
   removeFile(file) {
```
same as removeDestination

Tcharl replied:
Sorry, but I don't get the remark

Tcharl replied:
rolledbacked, but always sceptic

mshima commented on generators/base/generator-base-private.mjs:
```diff
@@ -142,9 +142,7 @@ export default class PrivateBase extends Generator {
    * Override yeoman generator's destinationPath to apply custom output dir.
    */
   destinationPath(...paths) {
```
dropped already

mshima commented on generators/base/generator-base.mjs:
```diff
@@ -203,12 +206,7 @@ export default class JHipsterBaseGenerator extends PrivateBase {
    * @returns {string}
    */
   jhipsterTemplatePath(...args) {
```
IMO the implementation should be keep here.

Tcharl replied:
I really want to move every single line of logic in dedicated file to see how it can be reused over the entire framework. I also want to strictly follow SOLID principles all over the business logic to maximize reuse.
The generator class (and all non-business logic related classes) will act as facade over this well OOP structured core logic files

mshima replied:
I cannot see another use of this extracted function.

mshima replied:
Can you think about any use of this?

Tcharl replied:
Yes...

The logic could be enhance to support when testing an home-made blueprint (part of the path is local, wether most of the paths are in the generator-jhipster package.

mshima replied:
I don't get why your are resolving questioning I am creating.
The generator is a class with inheritance, splitting into small utilities won't simplify the implementation.
I won't approved this.
Methods that belongs to the generator-base should be kept here, unless it's very complicated.

Tcharl replied:
Let take the example of code formatting: that are 6 different methods, each of them trimming, replacing specific characters, indenting, with a bunch of duplicated things. Suppose that we should change the behavior now: we'll modify all these methods, without knowing where they are, with the according amount of bugs, issues, etc.
Centralizing logic will help fixing this kind of issues.

Also, the current codebase is far from following the best code conventions (i.e. assignements + logic) and it's mostly due to this monolithic 2K+ lines of classes.

There are a lot of errors done with not centralizing these small methods somewhere.

On this specific use lines:
From a strict SOLID point of view, this method basically does two things: computing the root generator path, then fetching the path of a template.
While we for sure will only have one use case for locating template, we will for sure have many different kind of use case needing to get the root of the generator.
I'm also sure that it will help with typescript migration, feature addition, ...

Please don't be against this feature: I'm convinced that this will bring some new dimension to the framework, and most of the committers are for it, [at least deepu, sendil, pierre and me](https://groups.google.com/g/jhipster-dev/c/M1IKwib7i08/m/vV8DDnQ7AAAJ) and
https://groups.google.com/g/jhipster-dev/c/y_sER_CM3Hc/m/MhpXR3DaBAAJ

mshima replied:
Don't get me wrong, I am not denying any feature.
We probably have more than 100 methods at generator-base/private to extract.
`jhipsterTemplatePath`, `removeFolder`, `removeFile` implementation should be kept at the generator-base, at least for now.
From time to time I glance to see what to remove or move around, if the implementation is not local, I will have to open the other function.

Tcharl replied:
We could always mark them as private in the jsdoc, can't we?

Tcharl replied:
rolledback, but not convinced

mshima commented on generators/base/generator-base.mjs:
```diff
@@ -322,21 +312,7 @@ export default class JHipsterBaseGenerator extends PrivateBase {
    * @param {string} outputPath - Path to customize.
    */
   applyOutputPathCustomizer(outputPath) {
```
dropped already

Tcharl replied:
Done sir

mshima commented on generators/base/generator-base.mjs:
```diff
@@ -376,12 +348,7 @@ export default class JHipsterBaseGenerator extends PrivateBase {
    * @param {string} clientFramework - The name of the client framework
    */
   addIcon(iconName, clientFramework) {
```
My idea for needle has changed to use the source object.
We can inject the needle like here:
https://github.com/jhipster/generator-jhipster/blob/c6e0ea8035cb1fb2a9343d8067f73c6589e7622b/generators/docker/generator.mjs#L47-L74

and use like here:
https://github.com/jhipster/generator-jhipster/blob/c6e0ea8035cb1fb2a9343d8067f73c6589e7622b/generators/docker/generator.mjs#L100-L118

source is shared, so once injected every generator can use it.

mshima replied:
For now I would move the api to the client generator.

Tcharl replied:
Awesome, fully agree that this is the way to go.
Another idea out of my mind would be to be able to detect if the target code has been customized by the user pre merge. we could for example compare the fragment code that is in the 'jhipster_upgrade' branch with the content of the current branch, see if there's a diff.
If there's diff: don't merge but update the jhipster_upgrade branch with new code.
If there isn't: merge and update the baseline branch.

It would lead to a more frictionless custom + generated code cohabitation and a better dev XP in the end.

Nevertheless, the purpose of the merge request I currently provide is not to modify the behavior of the current codebase, but just to extract the logic in order to make it more simple to do it in a second round. I always prefer step-by-step approach rather than big bang changes (and I'm not enough confident doing so)

Tcharl replied:
Moved the api to the client generator

mshima commented on generators/base/generator-base.mjs:
```diff
@@ -395,12 +362,7 @@ export default class JHipsterBaseGenerator extends PrivateBase {
    * @param {string} translationKeyMenu - i18n key for entry in the menu
    */
   addElementToMenu(routerName, iconName, enableTranslation, clientFramework, translationKeyMenu = _.camelCase(routerName)) {
```
Move to the client generator.

Tcharl replied:
I will, for sure :-)

Tcharl replied:
I've tried, this one is quite hard to achieve (harder than it seems). I'll do it later in another round (and sorry for that)

Tcharl commented:
Would it be Ok to go, even with the current limitations?

mshima commented:
If we split methods marked to be removed, we will need to remove in 2 places later.

Tcharl commented:
I removed everything's not needed (or at least I think). Moving methods is 0 effort, but will drastically help future refactor

mshima commented on generators/client/logic/processors/build-tool/package-json.mjs:
```diff
@@ -0,0 +1,51 @@
+ * @param section the section of the package.json file where the dependency is
+ * @param keyToReplace the dependency kay to replace
+ */
+const substitutePackageJsonDependencyVersionAccordingToSource = (sourcePackageJson, targetPackageJson, section, keyToReplace) => {
```
The method that uses this was removed.

Tcharl replied:
removed

mshima commented on generators/entity/logic/enum.mjs:
```diff
@@ -0,0 +1,108 @@
+/**
```
Entity generator should be dumb, and is used only for adding/editing entities using prompts.
This should be placed at `generators/base-application` or `lib`.

Tcharl replied:
not fully agree: logic owned by a generator (here, enum is closer to entity than anything else) should be as close as possible to the one that is the most suited.
What are the argument against it?

Tcharl replied:
moved there

mshima commented on generators/base/generator-base.mjs:
```diff
@@ -225,66 +222,50 @@ export default class JHipsterBaseGenerator extends PrivateBase {
-      // React
-      // TODO:
-    }
+  isBuiltInUser(entityName) {
```
Those methods doesn't exists anymore at main

Tcharl replied:
removed, but please do not duplicate the tasks made here: rebasing is painful

mshima commented on generators/base/logic/generator/parsers/json-parser.mjs:
```diff
@@ -0,0 +1,25 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const parseJson = (filePath, defaultValue = {}) => {
```
The method that uses this was removed.

Tcharl replied:
Errata: removed the one related to package.json, I guess this particular one will be useful in some later

Tcharl commented:
@mshima starts to be a thing right?
could it be merged? I rebased 3 times, and the more I'll extract classes, the harder it will be

mshima commented:
@Tcharl I really appreciate your contribution almost all of my reviews were ignored.
- I continue to disagree that we should move generator related methods out of generator-base without a real purpose
- `isBuiltInEntity`, `isUsingBuiltInUser` were removed but are readied at this PR.
- I don't like the idea of having logics at `generators/entity`/`generators/entities` they are commands just like `generator/info`.
- `parseJson` should be removed, I don't agree with `Errata: removed the one related to package.json, I guess this particular one will be useful in some later`. A unused method is a left over and at least a code smell.

Others reviews:
- `generators/base/logic/asserts.mjs` is a utility of `generators/entity`. Should be moved there.
- `generators/base/logic/connect.mjs` maybe we should use `axios` which is already a dependency.
- I think it's far from been useful extract `const firstOrSecond = (first, second) => first || second;` we are not creating another declarative language, it's basic javascript `foo || bar` is much better than `firstOrSecond(foo, bar)`. By the way, we should use `foo ?? bar` for these cases now.

Tcharl commented:
> @Tcharl I really appreciate your contribution almost all of my reviews were ignored.

=> I swear that I did my best to take everything into account according to the current generator's constraints (remaining cjs is quite painful, it will need a lot of extract and refactor to definitely get rid of it)

> - I continue to disagree that we should move generator related methods out of generator-base without a real purpose

=> The purpose is always to introduce clear boundaries and maximize reuse across the business logic.

> - isBuiltInEntity, isUsingBuiltInUser were removed but are added again at this PR.

=> Sorry, rebase issue, fixed

> - I don't like the idea of having logics at generators/entity/generators/entities they are commands just like generator/info.

=> My idea here is to have a clear guideline and then being able to enforce them through archunit-like rules

> - parseJson should be removed, I don't agree with Errata: removed the one related to package.json, I guess this particular one will be useful in some later. A unused method is a left over and at least a code smell.

=> Done

Others reviews:

> - generators/base/logic/asserts.mjs is a utility of generators/entity. Should be moved there.

=> Done
> - generators/base/logic/connect.mjs maybe we should use axios which is already a dependency.

=> Not the purpose of this PR, can be done later on 
> - I think it's far from been useful extract const firstOrSecond = (first, second) => first || second; we are not creating another declarative language, it's basic javascript foo || bar is much better than firstOrSecond(foo, bar). By the way, we should use foo ?? bar for these cases now.

=> The logic behind was to centralize the context updates to be able to avoid it later on (looks evil to merge everything is `this`, even at the last step.

mshima commented on app.jdl:
```diff
@@ -0,0 +1,6 @@
+application {
```
drop

mshima commented on generators/base/generator-base-private.mjs:
```diff
@@ -148,11 +147,8 @@ export default class PrivateBase extends Generator {
-      this.log(`Removing the file - ${file}`);
-      shelljs.rm(file);
-    }
+    // TODO Should not update variable
+    file = deleteFile(this, file);
```
```suggestion
    file = this.destinationPath(file);
    this.log(`Removing the file - ${file}`);
    rmSync(file, { force: true });
```

Tcharl replied:
Not agree, extraction rule should be straightforward

Tcharl replied:
rollback, but always sceptic

mshima commented on generators/base/generator-base-private.mjs:
```diff
@@ -162,29 +158,8 @@ export default class PrivateBase extends Generator {
-      return !shelljs.exec(`git mv -f ${source} ${dest}`).code;
-    }
-    return true;
+    // TODO Should not update variable
+    folder = deleteFolder(this, folder);
```
```suggestion
    file = this.destinationPath(file);
    this.log(`Removing the folder - ${folder}`);
    rmSync(file, { force: true, recursive: true });
```

Tcharl replied:
Not agree, extraction rule should be straightforward

Tcharl replied:
taken into account

mshima commented on generators/base/generator-base-private.mjs:
```diff
@@ -465,40 +220,15 @@ export default class PrivateBase extends Generator {
    * @private
    * Check if Java is installed
    */
   checkJava() {
```
We don't need to have checkJava at base generator.

Tcharl replied:
I'll move it later: for now it's used by two generators so it's not that simple to extract.

Tcharl replied:
removed in favour of your implementation

mshima commented on generators/base/logic/connect.mjs:
```diff
@@ -0,0 +1,43 @@
+/**
```
I don't like the folder called `logic` I prefer `support`.
Please open a issue to discuss.

Tcharl replied:
support is less clear to me, still, the issue is here: https://github.com/jhipster/generator-jhipster/issues/20295

mshima commented on generators/base/logic/converter.mjs:
```diff
@@ -0,0 +1,41 @@
+ * @param {any} option - options
+ * @returns {boolean} true if option is in array and is set to 'true'
+ */
+const optionFromArray = (array, option) => {
```
Only used at server prompts. Should be an util there.

Tcharl replied:
I'm pretty sure that this could be useful for some generators. Also, the logic is not server-only

mshima commented on generators/base/logic/generator/locator.mjs:
```diff
@@ -0,0 +1,32 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const locateGenerator = (generator, env, options) => {
+  let existingGenerator;
+  try {
+    existingGenerator = generator || env.requireNamespace(options.namespace).generator;
+  } catch (error) {
+    if (options.namespace) {
+      const split = options.namespace.split(':', 2);
+      existingGenerator = split.length === 1 ? split[0] : split[1];
+    }
+  }
+  return existingGenerator;
+};
```
This is really confusing as a generator member, I think it's impossible to follow out of generator class context.

Tcharl replied:
I really don't want to have three different utils classes (`lib` folder, `utils.mjs, `logic`folder). which is why I added it there.
Also, it was originally part of `generator-base` so it does not harm more being here...

Tcharl replied:
rolledback

mshima commented on generators/base/logic/logging.mjs:
```diff
@@ -0,0 +1,76 @@
+  return (context.configOptions && context.configOptions.isDebugEnabled) || (context.options && context.options.debug);
+};
+
+/**
+ * Prints the messages then the arguments using the yeoman logger
+ * @param context the generator context.
+ * @param message message to print as a prefix
+ * @param args arguments to print
+ */
+const printMessageUsingGeneratorLogger = (context, message, ...args) => {
+  context.log(message);
+  args.forEach(arg => context.log(arg));
+};
+
+/**
+ * Checks if the configOptions or generatorOptions contains the debug flag.
+ * @param context the generator context.
+ */
+const generatorDebugOptionChecker = context => {
+  return context._debug && context._debug.enabled;
+};
+
+/**
+ * Print message using then internal yeoman debugger.
+ * @param context the generator context.
+ */
+const printUsingInternalDebugger = (context, message, ...args) => {
+  context._debug(message);
+  args.forEach(arg => context._debug(arg));
+};
+
+const logging = (context, message, ...args) => {
+  const formattedMsg = formatDebugMessageHeader(message);
+  if (jhipsterDebugOptionsChecker(context)) {
+    printMessageUsingGeneratorLogger(context, formattedMsg, ...args);
+  }
+  if (generatorDebugOptionChecker(context)) {
+    printUsingInternalDebugger(context, formattedMsg, ...args);
+  }
+};
```
Makes no sense to me.

Tcharl replied:
I just extracted the code, touched nothing here...
I'll also move info, warn and so on, then remove them from the generator file: it doesn't make sense to have it in a file which purpose is to orchestrate the generation

Tcharl replied:
I've the use case: jdl, cli and generator are using their own loggers (with the according redundancy and difference: cli one is more advanced than the other...).
Centralizing comes with value ^^. Will move it to cli, looks to make more sense there

Tcharl replied:
Makes sense to follow the refactoring later on: using the yeoman logger tight-couples any method/function that needs to log, even if it's only a stateless one to the yeoman generator, which is not a best practice at all.
I would proceed in two steps: extracting logging method, then use a logger class similar than the ones we find in the Java world (static logger like slf4j). Agrees?

mshima commented on generators/base/logic/options.mjs:
```diff
@@ -0,0 +1,38 @@
+ * @param {object} generator reference to the generator
+ * @param {object} options options
+ */
+const appendYeomanOptionsFromGeneratorOptions = (generator, options) => {
+  return {
+    root: options.root || generator.jhipsterTemplatesFolders || generator.templatePath(),
+    context: generator,
+    ...options,
+  };
+};
```
There is no reason for this.

Tcharl replied:
I just moved the code, I don't know if it's useful or not but it was there... I didn't want to introduce regresion. If not needed we'll remove it later

Tcharl replied:
always no idea of what it does: leaving it here for further investigation (will be easier now that I can debug with vscode, thanks to you^^)

mshima commented on generators/base/logic/output/file-operations.mjs:
```diff
@@ -0,0 +1,49 @@
+ */
+import shelljs from 'shelljs';
+
+const deleteFile = (context, file) => {
+  const destination = context.destinationPath(file);
+  if (destination && shelljs.test('-f', destination)) {
+    context.log(`Removing the file - ${destination}`);
+    shelljs.rm(destination);
+  }
+  return destination;
+};
+
+const deleteFolder = (context, folder) => {
+  const destination = context.destinationPath(folder);
+  if (destination && shelljs.test('-d', destination)) {
+    context.log(`Removing the folder - ${destination}`);
+    shelljs.rm('-rf', destination);
+  }
+  return destination;
+};
```
Should be dropped following reimplementation in other reviews.

Tcharl replied:
Not agree

Tcharl replied:
done

mshima commented on generators/base/logic/output/file-operations.mjs:
```diff
@@ -0,0 +1,49 @@
+  return destination;
+};
+
+const moveWithGit = (context, from, to) => {
+  const source = context.destinationPath(from);
+  const dest = context.destinationPath(to);
+  if (source && dest && shelljs.test('-f', source)) {
+    context.info(`Renaming the file - ${source} to ${dest}`);
+    return !shelljs.exec(`git mv -f ${source} ${dest}`).code;
+  }
+  return { source, dest };
+};
```
should be moved to generators/cleanup.js for now.
And dropped later.

Tcharl replied:
No!
The goal is to centralize the logic and to have each file with its own responsibility.

`logic` => Code
`cleanup` => list of file to remove
`templates` => template
`generator` => orchestration

Tcharl replied:
rolledback on base-private so that we know that it's debt. cleanup.js should stay clean without any other purpose than referencing files to cleanup

mshima commented on generators/base/logic/output/renderer.mjs:
```diff
@@ -0,0 +1,69 @@
+import { appendYeomanOptionsFromGeneratorOptions } from '../options.mjs';
+import { resetFakerSeed } from '../sequences.mjs';
+
+const appendCallBackToTemplate = (generator, cb, promise, source) => {
+  if (cb) {
+    return promise
+      .then(res => cb(res))
+      .catch(err => {
+        generator.warning(`Copying template ${source} failed. [${err}]`);
+        throw err;
+      });
+  }
+  return promise;
+};
```
??? why?
- it uses promises, await/async is much cleaner.

Tcharl replied:
I just extracted the code as-is.

Tcharl replied:
seems that this callback is always needed by many methods, leaving as-is for now

mshima commented on generators/base/logic/output/renderer.mjs:
```diff
@@ -0,0 +1,69 @@
+  return promise;
+};
+/**
+ * Render content
+ *
+ * @param {string} source source template
+ * @param {object} generator reference to the generator
+ * @param {any} context context
+ * @param {object} options options
+ * @param {function} [cb] callback function
+ * @return {Promise<String>} Promise rendered content
+ */
+const renderContent = (source, generator, context, options, cb) => {
+  const optionsWithYeoman = appendYeomanOptionsFromGeneratorOptions(generator, options);
+  if (source) {
+    const basename = path.basename(source);
+    resetFakerSeed(context, basename);
+  }
+  const promise = ejs.renderFile(generator.templatePath(source), context, optionsWithYeoman);
+  return appendCallBackToTemplate(generator, cb, promise, source);
+};
+
+const writeContent = (_this, context, destination, options, source) => {
+  // TODO simplify: looks like there's no need of so much params
+  return renderContent(source, _this, context, options)
+    .then(res => {
+      _this.fs.write(destination, res);
+      return destination;
+    })
+    .catch(error => {
+      this.warning(source);
+      throw error;
+    });
+};
```
will be dropped.

Tcharl replied:
I don't get it? Can it be dropped in subsequent PR?

mshima commented on generators/client/generator.mjs:
```diff
@@ -312,4 +313,29 @@ export default class JHipsterClientGenerator extends BaseApplicationGenerator {
   get [BaseApplicationGenerator.END]() {
     return this.asEndTaskGroup(this.delegateTasksToBlueprint(() => this.end));
   }
+
+  /**
+   * @private
+   * Add a new icon to icon imports.
+   *
+   * @param {string} iconName - The name of the Font Awesome icon.
+   * @param {string} clientFramework - The name of the client framework
+   */
+  addIcon(iconName, clientFramework) {
+    addIconInImport(this, iconName, clientFramework);
+  }
+
+  /**
+   * @private
+   * Add a new menu element, at the root of the menu.
+   *
+   * @param {string} routerName - The name of the Angular router that is added to the menu.
+   * @param {string} iconName - The name of the Font Awesome icon that will be displayed.
+   * @param {boolean} enableTranslation - If translations are enabled or not
+   * @param {string} clientFramework - The name of the client framework
+   * @param {string} translationKeyMenu - i18n key for entry in the menu
+   */
+  addElementToMenu(routerName, iconName, enableTranslation, clientFramework, translationKeyMenu = _.camelCase(routerName)) {
+    addMenuEntry(this, routerName, iconName, enableTranslation, clientFramework, translationKeyMenu);
+  }
```
Needles are client framework specific, and should change the parameters to 1 typed implementation

Tcharl replied:
Ok but for now I just moved the code as-is.

mshima commented on generators/client/logic/processors/needles.mjs:
```diff
@@ -0,0 +1,66 @@
+ * @param {string} iconName - The name of the Font Awesome icon.
+ * @param {string} clientFramework - The name of the client framework
+ */
+export const addIconInImport = (context, iconName, clientFramework) => {
+  if (clientFramework === ANGULAR) {
+    context.needleApi.clientAngular.addIcon(iconName);
+  } else if (clientFramework === REACT) {
+    // React
+    // TODO:
+  }
+};
+
+/**
+ * @private
+ * Add new entry to menu
+ * @param context the generator context
+ * @param routerName the client technology router name
+ * @param iconName name of the menu's icon
+ * @param enableTranslation whether translation is enabled
+ * @param clientFramework the client technology
+ * @param translationKeyMenu the translation key
+ */
+export const addMenuEntry = (
+  context,
+  routerName,
+  iconName,
+  enableTranslation,
+  clientFramework,
+  translationKeyMenu = _.camelCase(routerName)
+) => {
+  if (clientFramework === ANGULAR) {
+    context.needleApi.clientAngular.addElementToMenu(routerName, iconName, enableTranslation, translationKeyMenu, context.jhiPrefix);
+  } else if (clientFramework === REACT) {
+    // React
+    // TODO:
+  }
+};
```
They are client framework specific we should implement the parameter object type before moving around.

Tcharl replied:
Ok but for now I just moved the code as is. The purpose is this PR is not to modify the current logic

mshima commented on generators/entity/logic/asserts.mjs:
```diff
@@ -0,0 +1,66 @@
+ * @param isDecimal - flag indicating whether to check for decimal number or not
+ * @returns {number} parsed number if valid input; <code>NaN</code> otherwise
+ */
+const filterNumber = (input, isSigned, isDecimal) => {
+  const signed = isSigned ? '(\\-|\\+)?' : '';
+  const decimal = isDecimal ? '(\\.[0-9]+)?' : '';
+  const regex = new RegExp(`^${signed}([0-9]+${decimal})$`);
+
+  if (regex.test(input)) return Number(input);
+
+  return NaN;
+};
+
+/**
+ * @private
+ * @param {any} input input
+ * @returns {boolean} true if input is number; false otherwise
+ */
+const isNumber = input => {
+  return !isNaN(filterNumber(input));
+};
+
+/**
+ * @private
+ * @param {any} input input
+ * @returns {boolean} true if input is a signed number; false otherwise
+ */
+const isSignedNumber = input => {
+  return !isNaN(filterNumber(input, true));
+};
+
+/**
+ * @private
+ * @param {any} input input
+ * @returns {boolean} true if input is a signed decimal number; false otherwise
+ */
+const isSignedDecimalNumber = input => {
+  return !isNaN(filterNumber(input, true, true));
+};
+
```
👍 but I don't like the `logic` folder, IMO should be `support`.

mshima commented on generators/entities/logic/config/user.mjs:
```diff
@@ -0,0 +1,25 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import isUsingBuiltInCommonEntityConfiguration from './entities.mjs';
+
+const isBuiltInUserConfiguration = config => {
+  return isUsingBuiltInCommonEntityConfiguration(config);
+};
+
+export default isBuiltInUserConfiguration;
```
dropped

mshima commented on generators/entities/logic/index.mjs:
```diff
@@ -0,0 +1,20 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+export { default as isUsingBuiltInAuthorityConfiguration } from './config/authority.mjs';
+export { default as isBuiltInUserConfiguration } from './config/user.mjs';
```
dropped.

mshima commented on generators/entity/logic/entity/authority-checker.mjs:
```diff
@@ -0,0 +1,29 @@
+ * @param entityName the name of the entity to check.
+ * @returns {boolean} true if the entity is the built-in Authority entity.
+ */
+const entityIsAuthority = entityName => {
+  return _.upperFirst(entityName) === 'Authority';
+};
+export default entityIsAuthority;
```
dropped

mshima commented on generators/entity/logic/entity/user-checker.mjs:
```diff
@@ -0,0 +1,29 @@
+ * @param entityName the name of the entity to check.
+ * @returns {boolean} true if the entity is the built-in User entity.
+ */
+const entityIsUser = entityName => {
+  return _.upperFirst(entityName) === 'User';
+};
+export default entityIsUser;
```
dropped.

mshima commented on generators/entity/logic/enum.mjs:
```diff
@@ -0,0 +1,108 @@
+  };
+};
+
+export default enumInfo;
```
Enum is from jhipster base structure, `entity` generator is a command.
Should be moved to `base` or `lib`.

Tcharl replied:
done

mshima reviewed (changes requested):
Changes to `generator/base` should be written in typescript and implemented at `generator.mts`.

Tcharl commented:
> Changes to generator/base should be written in typescript and implemented at generator.mts.

I'm totally lost here: does it mean that this file will replace base-private and generator-base?
I'm also not confident at all achieving the migration to ts on the go: could it be done later on?

Tcharl commented:
@mshima can it be merged now? The more I progress adding new method extracted to this PR, the less I'm confident doing so. Also, rebasing is always painful.
I followed most of your remarks and can't really do more with the current scope: I'll for sure reopen some more until the overall framework will follow the guideline

Tcharl commented:
@mshima sorry to ask for a re-review, but I'm not comfortable adding more method extraction with enough confidence before having it merged: can you tell me the absolute minimum you want to be changed in order to get it merged please?

I also may have noticed an additional bug for microfrontend: the result of this test: needle-client-angular-generator.spec.mts expect a `module.then(...` if there's a microservice name, but last changes on the generator just generate a route instead.
I'll continue to investigate on that way, but as you were the contributor of both, you may know which one lies: the test or the needle...

Bests

Tcharl commented:
> Changes to `generator/base` should be written in typescript and implemented at `generator.mts`.

Can't it wait a bit? I'm mostly removing methods from there to facilitate further typescript rewrite later on: not so comfortable to achieve both in a row

mshima reviewed (changes requested):
My previous reviews seems to be ignored.
This a partial review.
I think the proposed implementation makes the implementation scattered and hard to follow.
The utilities can be implemented in typescript now that we don't have to old restrictions.

This PR is blocking others to contribute to the generator-base.

Tcharl commented on generators/app/support/asserts.mjs:
```diff
@@ -0,0 +1,43 @@
+
+import { packageJson } from '../../../lib/index.mjs';
+
+/**
+ * @private
+ * Check if Node is installed
+ */
+const checkNode = context => {
+  if (context.skipChecks) return;
+  const nodeFromPackageJson = packageJson.engines.node;
+  if (!semver.satisfies(process.version, nodeFromPackageJson)) {
+    context.warning(
+      `Your NodeJS version is too old (${process.version}). You should use at least NodeJS ${chalk.bold(nodeFromPackageJson)}`
+    );
+  }
+  if (!(process.release || {}).lts) {
+    context.warning(
+      'Your Node version is not LTS (Long Term Support), use it at your own risk! JHipster does not support non-LTS releases, so if you encounter a bug, please use a LTS version first.'
+    );
+  }
+};
```
I agree too, but it's hard to achieve in a row: if we don't split the methods first, most of the methods continue to be tied to generator context (i.e., the 'this.', 'context.', ...) which will make a bad api by design. It's why I thought to make it in two round. I know it's not perfect, and maybe comes with some 'cons' compared to the ideal way, but to me it's one way to achieve it iteratively.
I won't change this one yet, but will make my best to get the best trade off

Tcharl replied:
Done (not perfect, but still

Tcharl commented on generators/angular/support/needles.mjs:
```diff
@@ -0,0 +1,75 @@
+ * @param enableTranslation whether translation is enabled
+ * @param translationKeyMenu the translation key
+ */
+export const addMenuEntry = (context, routerName, iconName, enableTranslation, translationKeyMenu = _.camelCase(routerName)) => {
```
the angular needle need them right? Apart the context, which is needed by yeoman

Tcharl replied:
Angular or entity?
 - moving it close to entity generator will make it bound to angular,react and vue
 - moving it to angular will make these three bound to entities.

I may have made the bad choice (the second option) but anticipates the future typescript interface at the entity level and this one being an implementation

Tcharl replied:
will be tackled when we'll move to typescript (got this remark twice, for now it's just moving stuff and minimize contract change)

Tcharl commented on generators/base/support/connect.mjs:
```diff
@@ -0,0 +1,43 @@
+ *  @param {function} onSuccess - function, which gets called when the request succeeds, with the body of the response
+ *  @param {function} onFail - callback when the get failed.
+ */
+const httpsGet = (url, onSuccess, onFail) => {
+  https
+    .get(url, res => {
+      let body = '';
+      res.on('data', chunk => {
+        body += chunk;
+      });
+      res.on('end', () => {
+        onSuccess(body);
+      });
+    })
+    .on('error', onFail);
+};
```
I know, easier to refactor now that it is extracted right? The goal of this PR is not to change method signature, change technology or anything else, just to get rid of huge generator-base-private, generator-base and root/utils.mjs and spreading all these methods into their relative generator. This change is enough risky to avoid adding some more changes

Tcharl replied:
used axios

Tcharl commented on generators/base/support/context.mjs:
```diff
@@ -0,0 +1,23 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const generatorOrContext = (generator, context) => {
+  return generator || context;
+};
```
To identify where this bad pattern happens... The ugliest thing I've found in the codebase

Tcharl commented on generators/base/support/formatter.mjs:
```diff
@@ -0,0 +1,67 @@
+  return str === null || str === undefined || str.trim() === '';
+};
+
+const isSimpleText = (previousLine, text) => {
+  return !previousLine.endsWith('>') && !text.startsWith('<');
+};
```
Used in API javadoc and liquibase doc to identify if the code is custom or simple from what I guessed.

Tcharl replied:
doc added

Tcharl commented on generators/base/support/converter.mjs:
```diff
@@ -0,0 +1,41 @@
+ * @param {any} option - options
+ * @returns {boolean} true if option is in array and is set to 'true'
+ */
+const optionFromArray = (array, option) => {
+  let optionValue = false;
+  array.forEach(value => {
+    if (_.includes(value, option)) {
+      optionValue = value.split(':')[1];
+    }
+  });
+  optionValue = optionValue === 'true' ? true : optionValue;
+  return optionValue;
+};
```
server/support you mean ^^, no logic but in dedicated support classes.
Pretty sure we need it at many places and just duplicate this bits blindly.
Strange that, for example, it's not used for the client selection prompt/clientframeworkType.
Prefer to leave it there and try to reuse it at all needed places.

Tcharl commented on generators/base/generator-base-private.mjs:
```diff
@@ -162,563 +125,25 @@ export default class PrivateBase extends Generator {
+    const _context = context || _this;
+    await renderContent(source, _this, _context, options, res => {
+      callback(res);
     });
```
Which is why not iterating quickly on a PR is a nightmare: I would have reviewed this kind of adventure every two files modified, not that much more...

Tcharl commented on generators/angular/support/needles.mjs:
```diff
@@ -0,0 +1,75 @@
+ * @param {string} entityTranslationKeyMenu - i18n key for entity entry in menu
+ * @param {string} entityTranslationValue - i18n value for entity entry in menu
+ */
+export const addEntityMenuEntry = (context, routerName, enableTranslation, entityTranslationKeyMenu, entityTranslationValue, jhiPrefix) => {
```
see comments on addAdminMenuEntry.
Still, removed for further improvement later on.

Tcharl commented on generators/angular/support/needles.mjs:
```diff
@@ -0,0 +1,75 @@
+ * @param {boolean} enableTranslation - If translations are enabled or not
+ * @param {string} translationKeyMenu - i18n key for entry in the admin menu
+ */
+export const addAdminMenuEntry = (context, routerName, iconName, enableTranslation, translationKeyMenu) => {
```
This PR is not made to fix our clunky API contracts, just meant to extract them in dedicated files instead of these 2K+ base-private/base/utils files in order to ease further refactor.
I'm convinced that extracting methods will ease TS migration and method refactor in a second step.
Still, removed for further improvement later on.

Tcharl commented on generators/angular/support/needles.mjs:
```diff
@@ -0,0 +1,75 @@
+ */
+export const addIconInImport = (context, iconName) => {
+  context.needleApi.clientAngular.addIcon(iconName);
+};
```
The root cause is having the needleApi to be bound to it's 3 implementations doesn't make sense at all.
Neither the needle api to be loaded into the context.
Also, this refactor helper by not having a strict dependency between generator-base and angular generator during transition phase and being able to duplicate the same code within the 3 different clients.
Still, removed for further improvement later on.

Tcharl replied:
rolledback

Tcharl commented on generators/angular/generator.mjs:
```diff
@@ -118,4 +129,100 @@ export default class AngularGenerator extends BaseApplicationGenerator {
+   * @param {string} entityTranslationValue - i18n value for entity entry in menu
+   */
+  addEntityToMenu(
+    routerName,
+    enableTranslation,
+    entityTranslationKeyMenu = _.camelCase(routerName),
+    entityTranslationValue = _.startCase(routerName),
+    jhiPrefix = this.jhiPrefix
```
It's already one less compared to the previous version that was in base-private (the clientFramework param is no more here). The goal not being to change the entire signatures (yet)

Tcharl commented on generators/angular/generator.mjs:
```diff
@@ -118,4 +129,100 @@ export default class AngularGenerator extends BaseApplicationGenerator {
+   * @param {string} pageTitle - The translation key or the text for the page title in the browser
+   */
+  addEntityToModule(
+    entityInstance = this.entityInstance,
+    entityClass = this.entityClass,
+    entityName = this.entityAngularName,
+    entityFolderName = this.entityFolderName,
+    entityFileName = this.entityFileName,
+    entityUrl = this.entityUrl,
+    microserviceName = this.microserviceName,
+    readOnly = this.readOnly,
+    pageTitle = this.enableTranslation ? `${this.i18nKeyPrefix}.home.title` : this.entityClassPlural
```
It's already one less compared to the previous version that was in base-private (the clientFramework param is no more here). The goal not being to change the entire signatures (yet)

Tcharl commented on generators/angular/generator.mjs:
```diff
@@ -118,4 +129,100 @@ export default class AngularGenerator extends BaseApplicationGenerator {
+   * @param {boolean} enableTranslation - If translations are enabled or not
+   * @param {string} translationKeyMenu - i18n key for entry in the admin menu
+   */
+  addElementToAdminMenu(routerName, iconName, enableTranslation, translationKeyMenu = _.camelCase(routerName)) {
```
It's already one less compared to the previous version that was in base-private (the clientFramework param is no more here). The goal not being to change the entire signatures (yet)

Tcharl commented on generators/angular/generator.mjs:
```diff
@@ -118,4 +129,100 @@ export default class AngularGenerator extends BaseApplicationGenerator {
+    addAdminMenuEntry(this, routerName, iconName, enableTranslation, translationKeyMenu);
+  }
+
+  generateEntityClientEnumImports(fields) {
```
Generates the angular import section to of enums

DanielFran commented on generators/app/support/check-node.mjs:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/app/support/index.mjs:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/base/support/configuration-helpers/enum.mjs:
```diff
@@ -0,0 +1,108 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/base/support/configuration-helpers/options-helpers.mjs:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/base/support/configuration-helpers/options.mjs:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/base/support/configuration-helpers/sequences.mjs:
```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/base/support/generator-architecture/locator.mjs:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/base/support/index.mjs:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/base/support/logging.mjs:
```diff
@@ -0,0 +1,103 @@
+import chalk from 'chalk';
+
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/base/support/output/file-operations.mjs:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/base/support/output/renderer.mjs:
```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/base/support/remote.mjs:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/base/support/templates/doc-formatters.mjs:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/client/support/entity-definition.mjs:
```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/client/support/index.mjs:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/entity/support/index.mjs:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/entity/support/pk-utils.mjs:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/liquibase-changelogs/support/formatting.mjs:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/liquibase-changelogs/support/index.mjs:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/project-name/support/index.mjs:
```diff
@@ -0,0 +1,21 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/project-name/support/name-resolver.mjs:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/react/support/index.mjs:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/react/support/needles.mjs:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/vue/support/index.mjs:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

DanielFran commented on generators/vue/support/needles.mjs:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Copyright 2013-2022 the original author or authors from the JHipster project.
```
```suggestion
 * Copyright 2013-2023 the original author or authors from the JHipster project.
```

Tcharl commented on generators/angular/generator.mjs:
```diff
@@ -118,4 +129,100 @@ export default class AngularGenerator extends BaseApplicationGenerator {
+   * Check if language should be skipped for locale setting
+   * @param {string} language
+   */
+  skipLanguageForLocale(language) {
```
No idea, used here: `generators/angular/templates/src/main/webapp/app/app.module.ts.ejs`

Tcharl commented on generators/angular/generator.mjs:
```diff
@@ -118,4 +129,100 @@ export default class AngularGenerator extends BaseApplicationGenerator {
+    return out && out[0] && !!out[0].skipForLocale;
+  }
+
+  generateTestEntityId(primaryKey, index = 0, wrapped = true) {
```
used in angular templates (which is why it is here). just moved from base private...

Tcharl replied:
added doc

Tcharl commented on generators/angular/generator.mjs:
```diff
@@ -118,4 +129,100 @@ export default class AngularGenerator extends BaseApplicationGenerator {
+    return getTestEntityId(primaryKey, index, wrapped);
+  }
+
+  generateTestEntityPrimaryKey(primaryKey, index) {
```
doc added

Tcharl commented on generators/angular/generator.mjs:
```diff
@@ -118,4 +129,100 @@ export default class AngularGenerator extends BaseApplicationGenerator {
+    return getClientEnumImportsFormat(fields, ANGULAR);
+  }
+
+  getTypescriptKeyType(primaryKey) {
```
added doc

Tcharl commented on generators/base-docker/docker-base.mjs:
```diff
@@ -107,7 +97,7 @@ export function loadConfigs() {
 
   // Loading configs
-  this.debug(`Apps folders: ${this.appsFolders}`);
+  logDebug(this, `Apps folders: ${this.appsFolders}`);
```
The drawaback of the `this.anylog` is that it tight-couples any method that needs to log anything (even helpers, static, etc.) to the generator. the goal here is to make it in two steps: extracting everything to a central class, then reworking that class to cut the dependency to the generator

Tcharl commented on generators/base/generator-base-private.mjs:
```diff
@@ -162,563 +125,25 @@ export default class PrivateBase extends Generator {
-      return !shelljs.exec(`git mv -f ${source} ${dest}`).code;
-    }
-    return true;
-  }
```
rolledBack

Tcharl commented on generators/angular/generator.mjs:
```diff
@@ -118,4 +129,100 @@ export default class AngularGenerator extends BaseApplicationGenerator {
+   */
+  addIcon(iconName) {
+    addIconInImport(this, iconName);
+  }
```
removed

Tcharl commented:
I made a full pass on all the review messages, and it seems that all of them are resolved (rolled back, suggestion implemented, or out of scope due to disruptive contract change).
I hope it will be ok this time (to be able to contribute many other pr in this way, but smaller and easier to pass)
